debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbIncrementalSnapshotChangeEventSource.java

```diff
@@ -512,52 +521,49 @@ private void createDataEventsForDataCollection(MongoDbPartition partition) throw
-                chunkEndPredicate.put(DOCUMENT_ID, chunkEndOp);
-                predicate = new Document();
-                predicate.put("$and", Arrays.asList(chunkEndPredicate, maxKeyPredicate));
-            }
-
-            LOGGER.debug("\t For collection '{}' using query: '{}', key: '{}', maximum key: '{}'", currentCollection.id(),
-                    predicate.toJson(), context.chunkEndPosititon(), context.maximumKey().get());
+            Document predicate = constructQueryPredicate(context.chunkEndPosititon(), context.maximumKey().get());
+            LOGGER.debug("\t For collection '{}' using query: '{}', key: '{}', maximum key: '{}' to get all _id fields",
+                    currentCollection.id(), predicate.toJson(), context.chunkEndPosititon(), context.maximumKey().get());
 
             long rows = 0;
-            Timer logTimer = getTableScanLogTimer();
-
             Object[] lastRow = null;
             Object[] firstRow = null;
+            List<Future<?>> futureList = new ArrayList<>();
+            Object[] lastChunkKey = context.chunkEndPosititon();
 
-            for (BsonDocument doc : collection.find(predicate).sort(new Document(DOCUMENT_ID, 1))
-                    .limit(connectorConfig.getIncrementalSnapshotChunkSize())) {
+            for (BsonDocument doc : collection.find(predicate).sort(new Document(DOCUMENT_ID, 1)).projection(Projections.include(DOCUMENT_ID))
+                    .limit(chunkSize * threads)) {
                 rows++;
                 final Object[] row = new Object[]{ doc };
                 if (firstRow == null) {
                     firstRow = row;
                 }
-                final Struct keyStruct = currentCollection.keyFromDocumentSnapshot(doc);
-                window.put(keyStruct, row);
-                if (logTimer.expired()) {
-                    long stop = clock.currentTimeInMillis();
-                    LOGGER.debug("\t Exported {} records for collection '{}' after {}", rows, currentCollection.id(),
-                            Strings.duration(stop - exportStart));
-                    logTimer = getTableScanLogTimer();
+                lastRow = row;
+
+                if (rows % chunkSize == 0) {
+                    lastChunkKey = addChunkToExecutor(collection, lastRow, futureList, lastChunkKey);
                 }
+            }
 
-                lastRow = row;
+            // in case the last iteration doesn't have enough data, do it once again for the rest of the rows
+            if (rows % chunkSize != 0) {
+                addChunkToExecutor(collection, lastRow, futureList, lastChunkKey);
+            }
+
+            try {
+                for (Future<?> future : futureList) {
+                    future.get(); // Wait for the tasks to complete
+                }
+            }
+            catch (InterruptedException | ExecutionException e) {
+                throw new InterruptedException(e.toString());
```

Comment:
There is no need fo catching `InterruptedException` it can be propagagated directly from the method.
The `ExecutionException` should be handled like `throw new DebeziumException("Error while processing chunk", e)`

Reply:
done